### PR TITLE
Mobs for Nukkit 

### PR DESCRIPTION
**Pull request by PikyCZ**

### CODE REVIEW

**Comment by kazuemon** [comment]
nice :)

**Comment by NycuRO** [comment]
Eh, i don't like structure.

**Comment by Tee7even** [comment]
I'll try to check, test and, possibly, merge this today. Now I have to go...

**Comment by PikyCZ** [comment]
ok :D .. i create mob utils for upcoming new features ..

File: src/main/java/cn/nukkit/entity/monsters/EntitySkeleton.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        MobEquipmentPacket pk = new MobEquipmentPacket();
+        pk.eid = this.getId();
+        pk.item = new ItemBow();
+        pk.slot = 10;
+        pk.selectedSlot = 10;
+        player.dataPacket(pk);
+    }
+
+    @Override
+    public boolean entityBaseTick(int tickDiff) {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/monsters/EntitySkeleton.java]: You could make some kind of `BurningMob` abstract class for that behavior. Anyway, as I already said, it's early to implement it.

File: src/main/java/cn/nukkit/entity/monsters/EntityStray.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import cn.nukkit.level.Level;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.network.protocol.MobEquipmentPacket;
+import cn.nukkit.entity.mobutils.Drops;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class EntityStray extends EntityCreature {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/monsters/EntityStray.java]: IS-A skeleton.

File: src/main/java/cn/nukkit/entity/monsters/EntityZombieVillager.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class EntityZombieVillager extends EntityCreature {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/monsters/EntityZombieVillager.java]: IS-A zombie.

File: src/main/java/cn/nukkit/entity/passive/EntityCow.java [hunk 4]
```diff
@@ -41,10 +46,21 @@ public float getEyeHeight() {
     public String getName() {
         return this.getNameTag();
     }
-
+    
     @Override
     public Item[] getDrops() {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityCow.java]: http://minecraft.gamepedia.com/Cow#Drops

File: src/main/java/cn/nukkit/entity/passive/EntityDonkey.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package cn.nukkit.entity.passive;
+
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+public class EntityDonkey extends EntityAnimal {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityDonkey.java]: IS-A horse.

File: src/main/java/cn/nukkit/entity/passive/EntityMooshroom.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+public class EntityMooshroom extends EntityAnimal {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityMooshroom.java]: IS-A cow.

File: src/main/java/cn/nukkit/entity/passive/EntityMooshroom.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Override
+    public void initEntity() {
+        super.initEntity();
+        this.setMaxHealth(10);
+    }
+
+    @Override
+    public Item[] getDrops() {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityMooshroom.java]: Why didn't you implement such behavior for ordinary cows?

File: src/main/java/cn/nukkit/entity/passive/EntityMule.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package cn.nukkit.entity.passive;
+
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+public class EntityMule extends EntityAnimal {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityMule.java]: IS-A horse.

File: src/main/java/cn/nukkit/entity/passive/EntityPig.java [hunk 9]
```diff
@@ -44,7 +49,14 @@ public String getName() {
 
     @Override
     public Item[] getDrops() {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityPig.java]: http://minecraft.gamepedia.com/Pig#Drops

File: src/main/java/cn/nukkit/entity/passive/EntityRabbit.java [hunk 10]
```diff
@@ -44,7 +49,22 @@ public String getName() {
 
     @Override
     public Item[] getDrops() {
-        return new Item[]{Item.get(Item.RAW_RABBIT), Item.get(Item.RABBIT_HIDE), Item.get(Item.RABBIT_FOOT)};
+        List<Item> drops = new ArrayList<>();
+        if (this.lastDamageCause instanceof EntityDamageByEntityEvent) {
+            int rawrabbit = Drops.rand(1, 3); // drops 1-2 raws
+            int rabbithide = Drops.rand(1, 3); // drops 1-2 hide
+            int rabbitfoot = Drops.rand(0, 101) <= 9 ? 1 : 0; // with a 8,5% chance a rabbit foot is dropped
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityRabbit.java]: 10%?

File: src/main/java/cn/nukkit/entity/passive/EntitySkeletonHorse.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
+package cn.nukkit.entity.passive;
+
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+public class EntitySkeletonHorse extends EntityAnimal {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntitySkeletonHorse.java]: IS-A horse.

File: src/main/java/cn/nukkit/entity/passive/EntityZombieHorse.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
+package cn.nukkit.entity.passive;
+
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+public class EntityZombieHorse extends EntityAnimal {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/passive/EntityZombieHorse.java]: IS-A horse.

File: src/main/java/cn/nukkit/entity/monsters/EntityCaveSpider.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import cn.nukkit.event.entity.EntityDamageByEntityEvent;
+import cn.nukkit.item.Item;
+import cn.nukkit.level.format.FullChunk;
+import cn.nukkit.nbt.tag.CompoundTag;
+import cn.nukkit.entity.mobutils.Drops;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class EntityCaveSpider extends EntityCreature {
```
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/monsters/EntityCaveSpider.java]: IS-A spider. 

P.S. Not just `Creature`, but a mob/monster.

File: src/main/java/cn/nukkit/entity/mobutils/Drops.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
+package cn.nukkit.entity.mobutils;
+
+import java.util.Random;
+
+
+public class Drops {
```
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/mobutils/Drops.java]: This folder is for extends for mobs/monsters

File: src/main/java/cn/nukkit/entity/monsters/EntityBat.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+package cn.nukkit.entity.monsters;
```
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/monsters/EntityBat.java]: import creeper to monsters

File: src/main/java/cn/nukkit/entity/monsters/EntityEnderman.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    protected void initEntity() {
+        this.setMaxHealth(40);
+        super.initEntity();
+    }
+
+    @Override
+    public Item[] getDrops() {
+        List<Item> drops = new ArrayList<>();
+        if (this.lastDamageCause instanceof EntityDamageByEntityEvent) {
+            int EnderPearl = Drops.rand(0, 3); // drops 0-2 Pearls
```
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/monsters/EntityEnderman.java]: 0,2 = 0-1 
0,3= 0-2

File: src/main/java/cn/nukkit/entity/mobutils/Drops.java (hunk 14 above)
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/mobutils/Drops.java]: A Drops(Random) is for mobs

File: src/main/java/cn/nukkit/entity/passive/EntityPig.java (hunk 9 above)
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/passive/EntityPig.java]: drops is ok

File: src/main/java/cn/nukkit/entity/passive/EntityCow.java (hunk 4 above)
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/passive/EntityCow.java]: Drops is now ok

File: src/main/java/cn/nukkit/entity/monsters/EntitySkeleton.java (hunk 1 above)
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/monsters/EntitySkeleton.java]: early? .. nope its time to add :D

File: src/main/java/cn/nukkit/entity/monsters/EntityPigZombie.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        MobEquipmentPacket pk = new MobEquipmentPacket();
+        pk.eid = this.getId();
+        pk.item = new ItemSwordGold();
+        pk.slot = 10;
+        pk.selectedSlot = 10;
+        player.dataPacket(pk);
+    }
+
+    @Override
+    public Item[] getDrops() {
```
Comment (PikyCZ) [inline src/main/java/cn/nukkit/entity/monsters/EntityPigZombie.java]: Drops now is ok

File: src/main/java/cn/nukkit/entity/mobutils/Drops.java (hunk 14 above)
Comment (Tee7even) [inline src/main/java/cn/nukkit/entity/mobutils/Drops.java]: Bad design.

**Comment by Snake1999** [comment]
I'm so sorry but I have to say that, this kind of design is not what Nukkit community expected. Do not mix anything else, good or not, into this bad thing and daydreaming about we would accept this work.

**Pull request closed without merging**